src/ser/helpers.rs

```diff
@@ -1,19 +1,322 @@
-use serde::ser::Serialize;
-use ser::to_string;
+use serde::ser::{Impossible, Serialize, Serializer, Error as SerError};
+use error::Error;
 
-/// In XML, various things can only be serialized if they are (or aren't)
-/// primitive types. This is a **hack** to try and detect this.
-pub fn is_primitive<S>(thing: &S) -> bool
+
+/// Check if a value would be "wrapped" when serialized as XML. 
+///
+/// As a rule of thumb, if something **isn't** wrapped then the XML serialized
+/// version would look similar to what you get from `Display`. Wrapped things 
+/// typically look like `<foo>...</foo>`.
+pub fn is_wrapped<S>(thing: &S) -> bool
 where
     S: Serialize + ?Sized,
 {
-    let contains_xml_characters = match to_string(&thing) {
-        Ok(repr) => repr.contains("<") && repr.contains(">") && repr.contains("</"),
+    let mut is_wrapped = false;
+
+    thing.serialize(WrapSafeDetector{ is_wrapped: &mut is_wrapped }).ok();
+
+    is_wrapped
+}
+
+#[derive(Debug)]
+struct WrapSafeDetector<'a> {
+    is_wrapped: &'a mut bool,
+}
+
+#[allow(unused_variables)]
+impl<'a> Serializer for WrapSafeDetector<'a> {
+    type Ok = ();
```

Comment:
That doesn't fix the issue though. To return an `Ok` the type signatures say your `Ok` must be of type `Ok(Self::SerializeSeq)`. So you still need to create some struct with the 7 impls for `SerializeSeq`, `SerializeTuple`, `SerializeMap`, etc...

What benefit do you think removing the state would have?